### PR TITLE
[TE] Ring-of-experts TE permutation perf improvements

### PR DESCRIPTION
**Pull request by tdophung**
- Pass num_experts_per_shard to TE permute to mask routing to local experts
- Skip pre_bias_logits all_gather on TE path (wasted collective)
- Remove redundant input masking after group_sizes truncation
- Only apply post-matmul output mask on MT path, not TE
- Fix unpermute padding guard: use pad_offsets is not None
- Lower TE GMM alignment requirement from 128 to 8 (this is due to BF16, but can be changed to an enum or something)

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by jberchtold-nvidia — APPROVED** [review]
LGTM, thanks!
